aws/resource_aws_wafv2_web_acl.go

```diff
@@ -0,0 +1,843 @@
+	conn := meta.(*AWSClient).wafv2conn
+	ignoreTagsConfig := meta.(*AWSClient).IgnoreTagsConfig
+
+	params := &wafv2.GetWebACLInput{
+		Id:    aws.String(d.Id()),
+		Name:  aws.String(d.Get("name").(string)),
+		Scope: aws.String(d.Get("scope").(string)),
+	}
+
+	resp, err := conn.GetWebACL(params)
+	if err != nil {
+		if isAWSErr(err, wafv2.ErrCodeWAFNonexistentItemException, "") {
+			log.Printf("[WARN] WAFv2 WebACL (%s) not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	if resp == nil || resp.WebACL == nil {
+		return fmt.Errorf("Error getting WAFv2 WebACL")
+	}
+
+	d.Set("name", aws.StringValue(resp.WebACL.Name))
+	d.Set("capacity", aws.Int64Value(resp.WebACL.Capacity))
+	d.Set("description", aws.StringValue(resp.WebACL.Description))
+	d.Set("arn", aws.StringValue(resp.WebACL.ARN))
+	d.Set("lock_token", aws.StringValue(resp.LockToken))
+
+	if err := d.Set("default_action", flattenWafv2DefaultAction(resp.WebACL.DefaultAction)); err != nil {
+		return fmt.Errorf("Error setting default_action: %s", err)
+	}
+
+	if err := d.Set("rule", flattenWafv2WebACLRules(resp.WebACL.Rules)); err != nil {
+		return fmt.Errorf("Error setting rule: %s", err)
+	}
+
+	if err := d.Set("visibility_config", flattenWafv2VisibilityConfig(resp.WebACL.VisibilityConfig)); err != nil {
+		return fmt.Errorf("Error setting visibility_config: %s", err)
+	}
+
+	arn := aws.StringValue(resp.WebACL.ARN)
+	tags, err := keyvaluetags.Wafv2ListTags(conn, arn)
+	if err != nil {
+		return fmt.Errorf("Error listing tags for WAFv2 WebACL (%s): %s", arn, err)
+	}
+
+	if err := d.Set("tags", tags.IgnoreAws().IgnoreConfig(ignoreTagsConfig).Map()); err != nil {
+		return fmt.Errorf("Error setting tags: %s", err)
+	}
+
+	return nil
+}
+
+func resourceAwsWafv2WebACLUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafv2conn
+
+	log.Printf("[INFO] Updating WAFv2 WebACL %s", d.Id())
+
+	u := &wafv2.UpdateWebACLInput{
+		Id:               aws.String(d.Id()),
+		Name:             aws.String(d.Get("name").(string)),
+		Scope:            aws.String(d.Get("scope").(string)),
+		LockToken:        aws.String(d.Get("lock_token").(string)),
+		DefaultAction:    expandWafv2DefaultAction(d.Get("default_action").([]interface{})),
+		Rules:            expandWafv2Rules(d.Get("rule").(*schema.Set).List()),
+		VisibilityConfig: expandWafv2VisibilityConfig(d.Get("visibility_config").([]interface{})),
+	}
+
+	if v, ok := d.GetOk("description"); ok && len(v.(string)) > 0 {
+		u.Description = aws.String(v.(string))
+	}
+
+	err := resource.Retry(5*time.Minute, func() *resource.RetryError {
+		_, err := conn.UpdateWebACL(u)
+		if err != nil {
+			if isAWSErr(err, wafv2.ErrCodeWAFUnavailableEntityException, "") {
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if isResourceTimeoutError(err) {
+		_, err = conn.UpdateWebACL(u)
+	}
+
+	if err != nil {
```

Comment:
if the error here is a `WAFOptimisticLockException`, AWS docs recommend: 
```
Get the resource again, make any changes you need to make to the new copy, and retry your operation.
```
so I'm wondering as a practitioner, is that apparent if the error were returned to you here? ie. run terraform `plan` to get the new `LockToken` into state and then `apply` again to make the updates, not just retry an `apply` of the plan already created. i say this to think if we need to provide some additional error logging

Reply:
@anGie44 it can probably be improved indeed, but if this is done through logging would this message be visible to the end-user? I thought logging should be enabled.

I'm open to any (code) suggestion.

Reply:
oh by "logging" i think we can follow the same `fmt.Errorf` convention but catch this specific case with:
```
	if isAWSErr(err, wafv2. WAFOptimisticLockException, "") {
            return fmt.Errorf("Error updating WAFv2 WebACL, resource has changed since last refresh: %w", err)
	}
       return fmt.Errorf("Error updating WAFv2 WebACL: %s", err)
```

Reply:
aaah ok, clear :) ... tried to make it as  clear as possible